src/celeritas/optical/surface/detail/InitBoundaryExecutor.hh

```
@@ -52,6 +52,10 @@ CELER_FUNCTION void InitBoundaryExecutor::operator()(CoreTrackView& track) const
                && sim.status() == TrackStatus::alive;
     }());
 
+#if CELERITAS_CORE_GEO == CELERITAS_CORE_GEO_VECGEOM
+    // \todo VecGeom surface normals are not currently supported, see PR #1926
+    track.sim().status(TrackStatus::killed);
+#else
```

We should probably emit a warning somewhere if we're running an optical problem with VecGeom. Would it be difficult to still allow the track to propagate through the geometry/undergo volumetric processes but ignore any surface physics?

---

Yes, I was thinking that in the constructor of the boundary *action class constructor*, we call `CELER_NOT_IMPLEMENTED(...)` with a link to the issue. That would disable optical physics entirely in practice, which is what we basically have to do.

---

Looking at some of the options to handle this:

1. Putting the absorption here allows volumetric processes and internal traversal but not passing through a boundary (this was the behavior before this PR). Can add a warning / error message to state VecGeom isn't currently supported.
2. Throwing in the class constructor means `SurfacePhysicsParams` will fail to build and all VecGeom tests involving surface physics will error out and need to be disabled / masked somehow.
3. Can not build surface physics, but then we will need to add some more behavior as an alternative for when a VecGeom track reaches a boundary.
4. Add a trivial surface interactor that doesn't depend on surface normals. I've started working on this but adding custom models isn't straightforward for surface physics right now.

Of course open to other ideas. I think 1 touches the least amount of code

---

Ah, I forgot we were already killing the track in the post-boundary action. I'd lean toward still being able to run with VecGeom if possible, though it is silly to kill the track at the first boundary. A trivial surface interactor is a nice idea (especially if implementing the normal in VecGeom is going to take a while) but probably not worth the effort right now if it's not straightforward to add.

---

Aha, good delineation of the consequences @hhollenb . I think as we transition to actually testing the whole optical transport loop, it's better to have a prompt detailed failure rather than silent wrong results after the run is complete. So I think it might be best to choose #2 and disable optical testing for vecgeom builds because, well, optical physics needs surface normals except in the infinite case. I can add this to your branch if it's onerous (but indeed I think we had it in place at one point).

---

Ok so I did some more testing, and if you replace the `InitBoundaryExecutor` with just crossing the boundary (no scattering, normal, etc), then the discrepancy still appears:
```
  CELER_FUNCTION void InitBoundaryExecutor::operator()(CoreTrackView& track) const
  {
       CELER_EXPECT([track] {
           auto sim = track.sim();
           return sim.post_step_action()
                      == track.surface_physics().scalars().init_boundary_action
                  && sim.status() == TrackStatus::alive;
       }());
   
       auto geo = track.geometry();
       CELER_EXPECT(geo.is_on_boundary());
  
       // Move the particle across the boundary
       geo.cross_boundary();
       if (CELER_UNLIKELY(geo.failed()))
       {   
           track.apply_errored();
           return;
       }   
  
       if (!track.material_record().material_id())
       {   
           track.sim().status(TrackStatus::killed);
           return;
       }   
  }
```
Getting a different number of steps for just g4 11.0.2 and the vecgeom+g4@11.0.2 builds, same as above. So it's not a vecgeom normal issue and not a surface physics specific issue I think?

(just to be sure, added `CELER_NOT_IMPLEMENTED` to the VecGeom normal function and it doesn't throw any errors)

---

Another idea (though out of scope for this PR): currently I think disabling the boundary process in `GeantOpticalPhysicsOptions` will disable the process in Geant4 but have no effect in Celeritas. We could support that option by adding another boundary executor that just crosses the boundary/kills the track if it's not in an optical material and build that boundary action instead of the surface physics ones if the boundary option is disabled.